### PR TITLE
Task/221 implement insert adjacent functions

### PR DESCRIPTION
**Pull request by dkocsis-emarsys**
Hi!

I've finished with the changes related to the insertAdjacent functionalities and the missing implementation if insertBefore's null parameter handling.
If there's anything i should change, just tell me.
As for the solution, i basically went with a polyfill like solution, because it looked just fine to add this new functionality without write huge chunks of code. That's why i had to fix insertBefore, it depends on that missing behaviour.

### CODE REVIEW

File: packages/happy-dom/src/nodes/node/INode.ts [hunk 1]
```diff
@@ -21,7 +22,10 @@ export default interface INode extends IEventTarget {
 	cloneNode(deep: boolean): INode;
 	appendChild(node: INode): INode;
 	removeChild(node: INode): void;
-	insertBefore(newNode: INode, referenceNode?: INode): INode;
+	insertBefore(newNode: INode, referenceNode?: INode | null): INode;
+	insertAdjacentElement(position: TInsertAdjacentPositions, node: INode): INode | null;
```
Comment (capricorn86) [inline packages/happy-dom/src/nodes/node/INode.ts:26]: insertAdjacentElement(), insertAdjacentHTML(), insertAdjacentText() is part of the IElement interface and should be implemented in the Element class.

File: packages/happy-dom/src/nodes/node/Node.ts [hunk 2]
```diff
@@ -351,6 +359,60 @@ export default class Node extends EventTarget implements INode {
...
 	}
 
+	/**
+	 * Inserts a node to the given position.
+	 *
+	 * @param position Position to insert element.
+	 * @param element Node to insert.
+	 * @return Inserted node or null if couldn't insert.
+	 */
+	public insertAdjacentElement(position: TInsertAdjacentPositions, element: INode): INode | null {
```
Comment (capricorn86) [inline packages/happy-dom/src/nodes/node/Node.ts:369]: insertAdjacentElement(), insertAdjacentHTML(), insertAdjacentText() is part of the IElement interface and should be implemented in the Element class.

File: packages/happy-dom/src/nodes/node/Node.ts [hunk 3]
```diff
@@ -351,6 +359,60 @@ export default class Node extends EventTarget implements INode {
...
+	 * Inserts an HTML string to the given position.
+	 *
+	 * @param position Position to insert text.
+	 * @param text HTML string to insert.
+	 * @return Inserted node or null if couldn't insert.
+	 */
+	public insertAdjacentHTML(position: TInsertAdjacentPositions, text: string): void {
+		const templateElement = <IHTMLTemplateElement>this.ownerDocument.createElement('template');
+		templateElement.innerHTML = text;
+		this.insertAdjacentElement(position, templateElement.content);
```
Comment (capricorn86) [inline packages/happy-dom/src/nodes/node/Node.ts:401]:
```suggestion
		for(const node of XMLParser.parse(text).childNodes.slice()) {
		      this.insertAdjacentElement(position, node);
		}
```
I think we can use the XMLParser directly here :slightly_smiling_face: 

Note that you need to prepend the commit with "#228@minor:" if you apply the suggestion as well.

**Review by capricorn86 — CHANGES_REQUESTED** [review]
Looks really good and thank you for contributing! :+1: 

I just have some minor comments :slightly_smiling_face:

**Comment by capricorn86** [comment]
The tests where really good by the way :+1:

**Comment by dkocsis-emarsys** [comment]
Thanks for the review, I've made the requested changes.
I've used #221 for moving the things to the Element class and #228 for the suggestion.

**Review by capricorn86 — APPROVED** [review]
Awesome work! :+1:

**Pull request merged**